habitat_sim/utils/data/dataextractor.py

```diff
@@ -0,0 +1,283 @@
+        x, z = self.topdown_view.shape
+        self.dist = (
+            min(x, z) // 10
+        )  # This produces lots of simular images for small scenes. Perhaps a smarter solution exists
+
+    def extract_poses(self, labels):
+        height, width = self.topdown_view.shape
+        n_gridpoints_width, n_gridpoints_height = (
+            width // self.dist - 1,
+            height // self.dist - 1,
+        )
+        self.gridpoints = []
+        for h in range(n_gridpoints_height):
+            for w in range(n_gridpoints_width):
+                point = (self.dist + h * self.dist, self.dist + w * self.dist)
+                if self.valid_point(*point):
+                    self.gridpoints.append(point)
+
+        # Find the closest point of the target class to each gridpoint
+        poses = []
+        self.cpis = []
+        for point in self.gridpoints:
+            closest_point_of_interest, label = self._bfs(point, labels)
+            if closest_point_of_interest is None:
+                continue
+
+            poses.append((point, closest_point_of_interest, label))
+            self.cpis.append(closest_point_of_interest)
+
+        # Convert from topdown map coordinate system to that of the pathfinder
+        startw, starty, starth = self._get_pathfinder_reference_point()
+        for i, pose in enumerate(poses):
+            pos, cpi, label = pose
+            r1, c1 = pos
+            r2, c2 = cpi
+            new_pos = np.array([startw + c1 * self.res, starty, starth + r1 * self.res])
+            new_cpi = np.array([startw + c2 * self.res, starty, starth + r2 * self.res])
+            cam_normal = new_cpi - new_pos
+            new_rot = self._compute_quat(cam_normal)
+            poses[i] = (new_pos, new_rot, label)
+
+        return poses
+
+    def valid_point(self, row, col):
+        return self.topdown_view[row][col] == 1.0
+
+    def is_point_of_interest(self, point, labels):
+        r, c = point
+        is_interesting = False
+        if self.topdown_view[r][c] in labels:
+            is_interesting = True
+
+        return is_interesting, self.topdown_view[r][c]
+
+    def _show_topdown_view(self, cmap="seismic_r", show_valid_points=False):
+        if show_valid_points:
+            topdown_view_copy = copy.copy(self.topdown_view)
+            for p in self.gridpoints:
+                r, c = p
+                topdown_view_copy[r][c] = 0.5
+
+            for cpi in self.cpis:
+                r, c = cpi
+                topdown_view_copy[r][c] = 0.65
+
+            plt.imshow(topdown_view_copy, cmap=cmap)
+        else:
+            plt.imshow(self.topdown_view, cmap=cmap)
+
+        plt.show()
+
+    def _get_pathfinder_reference_point(self):
+        bound1, bound2 = self.pathfinder.get_bounds()
+        startw = min(bound1[0], bound2[0])
+        starth = min(bound1[2], bound2[2])
+        starty = self.pathfinder.get_random_navigable_point()[
```

Comment:
If you have the caller send in the height for generating the top-down map, then you can get this exactly.